### ISSUE TITLE
Module not found: Error

### ISSUE THREAD

**Issue by mbouclas** [dherges/ng-packagr#117]
## Type of Issue

- [x] Bug Report
- [ ] Feature Request

## Description

After trying it on a dead simple app and failed, i cloned the example repo, ng-packaged, and got the same exact error. It builds without a problem, but the moment you try to import the lib to the app you get the error
```
ERROR in ./src/app/app.module.ts
Module not found: Error: Can't resolve '@my/lib' in 'D:\Work\testing\Angular2\ng-packaged\src\app'
resolve '@my/lib' in 'D:\Work\testing\Angular2\ng-packaged\src\app'
  Parsed request is a module
  using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src/app)
    Field 'browser' doesn't contain a valid alias configuration
  after using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src/app)
    resolve as module
      D:\Work\testing\Angular2\ng-packaged\src\app\node_modules doesn't exist or is not a directory
      D:\Work\testing\Angular2\ng-packaged\src\node_modules doesn't exist or is not a directory
      D:\Work\testing\Angular2\node_modules doesn't exist or is not a directory
      D:\Work\node_modules doesn't exist or is not a directory
      D:\node_modules doesn't exist or is not a directory
      looking for modules in D:\Work\testing\Angular2\ng-packaged\node_modules
        using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./node_modules)
          using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./node_modules/@my/lib)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib.js doesn't exist
            as directory
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib doesn't exist
      looking for modules in D:\Work\testing\node_modules
        No description file found
        Field 'browser' doesn't contain a valid alias configuration
        No description file found
        no extension
          Field 'browser' doesn't contain a valid alias configuration
          D:\Work\testing\node_modules\@my\lib doesn't exist
        .ts
          Field 'browser' doesn't contain a valid alias configuration
          D:\Work\testing\node_modules\@my\lib.ts doesn't exist
        .js
          Field 'browser' doesn't contain a valid alias configuration
          D:\Work\testing\node_modules\@my\lib.js doesn't exist
        as directory
          D:\Work\testing\node_modules\@my\lib doesn't exist
      looking for modules in D:\Work\testing\Angular2\ng-packaged\node_modules
        using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./node_modules)
          using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./node_modules/@my/lib)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib.js doesn't exist
            as directory
              D:\Work\testing\Angular2\ng-packaged\node_modules\@my\lib doesn't exist
      looking for modules in D:\Work\testing\Angular2\ng-packaged\src
        using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src)
          using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src/@my/lib)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib.js doesn't exist
            as directory
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib doesn't exist
      looking for modules in D:\Work\testing\Angular2\ng-packaged\src
        using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src)
          using description file: D:\Work\testing\Angular2\ng-packaged\package.json (relative path: ./src/@my/lib)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib.js doesn't exist
            as directory
              D:\Work\testing\Angular2\ng-packaged\src\@my\lib doesn't exist
```

The angular app.module.ts
```
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HttpModule } from '@angular/http';

import { AppComponent } from './app.component';
import { MyLibModule } from '@my/lib';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    MyLibModule.forRoot()
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
tsconfig.app.json
```
{
  "extends": "../tsconfig.json",
  "angularCompilerOptions": {
    "paths": {
      "@my/lib": [ "../dist/my-lib" ]
    }
  },
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "module": "es2015",
    "baseUrl": "",
    "types": [],
    "paths": {
      "@my/lib": [ "../dist/my-lib" ]
    }
  },
  "exclude": [
    "test.ts",
    "**/*.spec.ts"
  ]
}
```

main tsconfig.json
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "baseUrl": "src",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2016",
      "dom"
    ],
    "paths": {
      "@my/lib": [ "dist/my-lib" ]
    }
  },
  "exclude": [
    ".ng_build"
  ]
}

```

#### How To Reproduce

$ yarn install
$ yarn build:lib
$ ng serve

#### Version Information

```
ng-packagr: v1.0.1
node: v8.1.2
@angular: v4.3.2
rxjs:
zone.js:
```

**Comment by dherges** [comment]
Hi @mbouclas,

there is an issue with the `paths` mapping in the tsconfig. Builds [started failing with angular cli 1.3.x - issue angular-cli#7341](https://github.com/angular/angular-cli/issues/7341#issuecomment-324839657)

You can try to use a symlink instead. Something like:

`node_modules/@my/lib` --symlink to--> `dist/@my/lib`

In another project, a workaround is to copy the dist folder to the `src` folder angular cli, i.e.:
```
src
 |- app
 |- @my
     |- lib
 |- tsconfig.app.json (paths mapping to "./@my/lib")
```

That somehow works w/ angular cli 1.3.2.

**Comment by mbouclas** [comment]
got it, thanks.

**Comment by mbouclas** [comment]
@dherges when creating the symlink to node_modules it does seem to import it, but it produces the following error
```
Uncaught Error: Unexpected value 'MyModule' imported by the module 'AppModule'. Please add a @NgModule annotation.
```
Sure enough the exported module looks like this
```
export declare const COMPONENTS: any[];

export declare class MyModule {
    constructor();
    static forRoot(): {
        ngModule: typeof MyModule ;
        providers: any[];
    };
}
```

So, using the symlink trick, do we still have to add something to the app tsconfig or am i missing something?

**Comment by dherges** [comment]
No, with the symlink you shuld remove paths from tsconfig. So either tsconfig or symlink. Do not use both in parallel (don't know what happens then...)

Is the symlink in the correct direction? I always happen to exchange source and target. It needs to be like this:

```
|- dist
    |- @my
        |- lib
            |- package.json
|- node_modules
   |- @my
       |- lib
```

Now, when you do `cat node_modules/@my/lib/package.json` the file from the dist folder should show up. I just want to make sure no silly mistakes are in place :)

Or:
```bash
$ ls -al node_modules/@my
lib --> ../../dist/@my/lib
```

So the command for `ln` would/should be something like:

`ln -sf ../../dist/@my/lib node_modules/@my/lib`

**Comment by mbouclas** [comment]
Yes the symlink is correct and i do see the changes happen. Also, i have removed everything path related from tsconfig. The error produced suggests that it did find the package in node_modules, it just doesn't like  the way it's formatted causes it complains for a missing annotation `@NgModule`

Is there a chance that it is loading the wrong file? Also. noticed that this message shows up when importing the module plain, if i try to import it like `MyModule.forRoot()` i get another error 
```
vendor.bundle.js:29253 Uncaught Error: Unexpected value '[object Object]' imported by the module 'AppModule'. Please add a @NgModule annotation.
```

**Comment by mbouclas** [comment]
For anyone with a similar issue, my way of solving this was not to symlink (tested on windows only using both soft and hard links). For some reason this just won't work. What does work though is copying the dist output to your node_modules folder. 

I ended up using a chokidar watcher which builds the package and then copies the dist to node_modules.

@dherges thanks for your time

**Comment by MasonOh91** [comment]
@mbouclas your solution worked for me as well after a failed attempt at running a `yarn link` or `npm link`. Thanks for the heads up on this.

**Comment by riccardomariani** [comment]
I'm getting the same type of error even not using "paths". 
I'm trying to create a library from a module inside a regular angular-cli project.
I followed the guide here: https://medium.com/@ngl817/building-an-angular-4-component-library-with-the-angular-cli-and-ng-packagr-53b2ade0701e but I'm getting the error reported in this issue.
You can find my test in this github project: https://github.com/riccardomariani/odata-v4-ng.
Any suggestion to solve the problem?
Thanks a lot in advance

**Comment by dherges** [comment]
Can you take a look at [copying the dist artefacts](https://github.com/dherges/ng-packaged/commit/7d0c995e173062921fc6c57a1c4444b508188e5f) and [mapping paths to sources for a dev build](https://github.com/dherges/ng-packaged/commit/992d643e8bc5cd90ea24d4574868bc320be64039)

These solve the issues for TypeScript module resolution that came with angular/angular-cli#7341 - see https://github.com/angular/angular-cli/issues/7341#issuecomment-324839657 

Regarding `Unexpected value '[object Object]' imported ... Please add a @NgModule annotation.`, may have another cause. Please check and/or try to debug whether the import is properly resolved!

**Comment by riccardomariani** [comment]
Actually the problem was caused by how I was importing classes from the library.
Based on the generated library code, Visual Studio Code proposed
import { ODataModule } from 'odata-v4-ng/odata-v4-ng'; or
import { ODataModule } from 'odata-v4-ng/public_api';
instead of
import { ODataModule } from 'odata-v4-ng';
Changing the name of the file generated by ng-packagr from odata-v4-ng.d.ts to index.d.ts solved the problem and now Visual Studio Code is proposing the right import.
So my question is: is this a misconfiguration of ng_packagr in my library? or is this the expected behavior?

**Comment by riccardomariani** [comment]
I changed ng-package.json config file adding:     "flatModuleFile": "index" 
and this solved the problem.

{
    "$schema": "./node_modules/ng-packagr/ng-package.schema.json",
    "lib": {
        "entryFile": "public_api.ts",
        "flatModuleFile": "index"
    }
}

**Comment by dherges** [comment]
There are several ways to work with a library (and it's `dist` artefacts) and Angular `ng` CLI in the same repo:

- copying `dist` artefacts: demonstrated by `build:app` in [ng-packaged demo](https://github.com/dherges/ng-packaged/blob/master/package.json#L9-L10)
- using tsconfig paths mapping: demonstrated by `build:dev` in [ng-packaged demo](https://github.com/dherges/ng-packaged/blob/master/package.json#L9-L10)
- symlink the `dist` folder (also `yarn link` for creating/managing the symlink), as in the [in the consumer integration test](https://github.com/dherges/ng-packagr/tree/master/integration#sample-libraries).

**Comment by sumigoma** [comment]
@dherges I have tried all of these options, but the slight difference is that I intend on ultimately publishing my updated library to npm, so I'm trying to use `npm link` to replace the version in `node_modules`, and I do not want it to be a monorepo with the library in `src`.

My error is very similar to OP of this issue, but I _have_ tried copying the dist artifacts (from a different project) to node_modules (with no path mapping) and I've tried doing the tsconfig path mapping with `npm link` and `ng build --aot` with `--preserve-symlinks`. They all seem to produce the same error, but only with AoT builds. 

As others have stated, `ng serve` works fine.
